packages/flutter/lib/src/gestures/tap.dart

```diff
@@ -98,6 +98,197 @@ typedef GestureTapCallback = void Function();
 typedef GestureTapCancelCallback = void Function();
 
+/// The base class of gesture recognizers that recognize taps.
+///
+/// Gesture recognizers take part in gesture arenas to enable potential gestures
+/// to be disambiguated from each other. This process is managed by a
+/// [GestureArenaManager].
+///
+/// A tap is defined as a sequence of events that starts with a down, followed
+/// by optional moves, then ends with an up. All move events must contain the
+/// same `buttons` with the down event, and must not be too far from the initial
+/// position. The gesture is rejected on any violation, a cancel event, or
+/// if any other gesture wins the arena. It is accepted only when it is the last
+/// member of the arena.
+///
+/// [BaseTapGestureRecognizer] considers all the pointers involved in the pointer
+/// event sequence as contributing to one gesture. For this reason, extra
+/// pointer interactions during a tap sequence are not recognized as additional
+/// taps. For example, down-1, down-2, up-1, up-2 produces only one tap on up-1.
+///
+/// [BaseTapGestureRecognizer] can not be directly used, since it does not
+/// define which buttons to accept, or what to do when a tap happens. If you
+/// want to build a custom tap recognizer, extend this class by overriding
+/// [isPointerAllowed] and the handler methods.
+///
+/// See also:
+///
+///  * [TapGestureRecognizer], a ready-to-use tap recognizer that recognizes
+///    taps of the primary button and taps of the secondary button.
+///  * [ModalBarrier], a widget that uses a custom tap recognizer that accepts
+///    any buttons.
+abstract class BaseTapGestureRecognizer extends PrimaryPointerGestureRecognizer {
+  /// Creates a tap gesture recognizer.
+  BaseTapGestureRecognizer({ Object debugOwner, Duration deadline = kPressTimeout })
+    : super(deadline: deadline, debugOwner: debugOwner);
+
+  bool _sentTapDown = false;
+  bool _wonArenaForPrimaryPointer = false;
+
+  PointerDownEvent _down;
+  PointerUpEvent _up;
+
+  /// A pointer that might cause a tap has contacted the screen.
+  ///
+  /// This triggers after the down event, once a short timeout ([deadline]) has
+  /// elapsed, or once the gestures has won the arena, whichever comes first.
+  ///
+  /// The parameter `down` is the down event of the primary pointer that started
+  /// the tap sequence.
+  ///
+  /// If the gesture doesn't win the arena, [handleTapUp] is called next.
+  /// Otherwise, [handleTapCancel] is called next.
+  @protected
+  void handleTapDown({ PointerDownEvent down });
+
+  /// A pointer that will trigger a tap has stopped contacting the screen.
+  ///
+  /// This triggers on the up event, if the gesture wins the arena with it
+  /// or has previously won.
+  ///
+  /// The parameter `down` is the down event of the primary pointer that started
+  /// the tap sequence, and `up` is the up event that ended the tap sequence.
+  ///
+  /// If the gesture doesn't win the arena, [handleTapCancel] is called
```

Comment:
Please double-check with other documentation to make sure this is consistent: I believe the recognizer is the one winning the arena, not the gesture?

Reply:
Searched `win.*arena` throughout the project and I think we're not really being consistent. What wins the arena can be members (captains), recognizers, or pointers. (One of them even says gesture detectors win the arena, which is obviously wrong).

I think "gestures winning the arena" is clear and unambiguous, although I can change it to "If this recognizer doesn't win the arena" instead.

Reply:
Let's not add another thing to the list of things that can win the arena then. I think we should change this to recognizers winning the arena.